### ISSUE TITLE
Making GoogleAuthException public

### ISSUE THREAD

**Issue by julianSelser** [googleapis/google-auth-library-java#1122]
I'm using the [google ads java lib](https://github.com/googleads/google-ads-java) which has this project as a dependency.

its `GoogleAdsClient` take these project's `GoogleCredentials` and when trying to authenticate, it may throw [`com.google.auth.oauth2.GoogleAuthException`](https://github.com/googleapis/google-auth-library-java/blob/main/oauth2_http/java/com/google/auth/oauth2/GoogleAuthException.java) (or `OAuthException`)

If so, we would love to catch it like this:

```
try {
  UserCredentials credentials = ...; //=> GoogleCredentials

  GoogleAdsServiceClient client = GoogleAdsClient.newBuilder()
    .setCredentials(credentials)
    .setDeveloperToken(token)
    .build()
    .getLatestVersion()
    .createGoogleAdsServiceClient();

  doSomethingWith(client);
} catch(GoogleAuthException ex) {
  // handle the specific auth exception
} catch(PossiblyOtherException ex) {
 // handle possibly another exception
}
```

But we cant because `GoogleAuthException` is package private, would be nice to have it as `public` (along with the other exceptions that may bubble outside the package)

**Comment by julianSelser** [comment]
Hey @TimurSadykov thanks for assigning yourself! let me know if Im missing anything

**Comment by TimurSadykov** [comment]
Hi, thanks for raising the issue! 

Exceptions are package private intentionally. We don't want to have flexibility to change them in the future without breaking change concerns. 

At the same time, we understand that there could be valid use cases for public exceptions. Could you please describe scenarios, that you think you can handle if Exceptions are public? 

**A proposal:** rely on interfaces. It is way easier to maintain public interfaces. You can already catch GoogleAuthException by checking [Retryable](https://github.com/googleapis/google-auth-library-java/blob/main/credentials/java/com/google/auth/Retryable.java) interface. 

How about we do the same for OAuthException? Let's define the OAuthError interface in the credentials package.

**Comment by julianSelser** [comment]
Thanks for the response Timur! We cant really catch `Retryable` since its not an `Exception`, unless you mean checking for the class or something of the sort. I fail to see how to use it otherwise.

I like the proposal if it means to extend`Exception`, also maybe instead of `OAuthError` should be called `OAuthException` since `Error` is on the `Throwable` hierarchy and already has semantics, but then we are back at the beggining having a public `OAuthException`.

My intention was to convince you of making the exceptions public but I wasnt sure myself tbh, and Im probably not smart enough to make the point anyways, so I asked chat GPT and the results were interesting:

<img width="729" alt="image" src="https://user-images.githubusercontent.com/4140960/214701635-c2ae8436-8fde-47de-906d-ecb62cf76d57.png">

<img width="732" alt="image" src="https://user-images.githubusercontent.com/4140960/214701766-0505db8a-444d-4909-9519-0696324385a1.png">

**Comment by TimurSadykov** [comment]
@julianSelser Yes, more specifically I meant catching IOExceptions and checking for instances of Retryable or other interfaces. This makes more sense also because IOException is already a historical de-facto standard exception type and if you want to catch all auth exceptions, you have to catch IOExceptions anyways. 

The name OAuthException works better than OAuth2Error, but there is already an exception called OAuthException and java does not allow interface and exception with the same name. So we need something else, ideas welcome :)

**Comment by julianSelser** [comment]
@TimurSadykov I'd rather have the exceptions public instead of forcing users to do an `instanceof` (and maybe a cast) but I guess its better than not being able to catch it.

Is it ok if I redo my PR to accomodate the interface? Im wondering if we can have some `reason()` method we can have for the user to know what happened, seems like should also implement Retryable, what do you think?

**Comment by TimurSadykov** [comment]
@julianSelser Ideally, we want to follow the OAuth2 spec for errors. It has a field called errorDescription which is similar to what you are describing? 

I would suggest following the OAuthException implementation as a base for you interface. It should have what you need. Then we can make the existing OAuthException to extend the new interface. It already extends the Retryable via GoogleAuthException.  

Then we can throw the updated exception from our credential classes. 

Let me know if any questions.

**Comment by lqiu96** [comment]
Closing this issue due to inactivity. If there are any other concerns or reasons to for making `GoogleAuthException` public, we can re-open this issue. Thanks!

**Comment by quodlibetor** [comment]
> If there are any other concerns or reasons to for making GoogleAuthException public

Thanks for considering this. My use case: I’m building a service that monitors GCP, and I want to report auth misconfigurations to users instead of silently retrying forever. Right now, the only reliable signal is whether an exception implements `com.google.auth.Retryable` — but checking `(e instanceof Retryable)` to detect misconfigurations is awkward and semantically off.

Ideally, I'd like more expressive interfaces like:

```java
interface MalformedCredentials {} // structurally invalid
interface InvalidCredentials {}   // valid format, but no access
```

These would make it easier to handle errors clearly and intentionally. The fact that `CredentialFormatException` may also be thrown by `com.google.auth` but _cannot_ be checked for is another wrinkle preventing my from just checking for `IOException`. Making `GoogleAuthException` public could work if it consistently represents permissions errors.

These would make it easier to handle errors clearly and intentionally. One additional wrinkle: `CredentialFormatException` is another exception that signals a specific kind of misconfiguration, but it isn’t public and doesn't implement `Retryable` so I can’t check for it either. This further limits the usefulness of catching plain `IOException`.
